File: Roadmap/03 - ESTRUCTURAS DE DATOS/python/worlion.py
# List
my_list = [5, 3, 1, 4, 2]
print ("\n\n____ LISTAS____");

print(f" - Lista:\t\t\t{my_list}")
my_list.append(8)
print(f" - Lista (añado el 8):\t\t{my_list}")
my_list.remove(3)
print(f" - Lista (borro el 3):\t\t{my_list}")
my_list[2] = 99
print(f" - Lista (actualizo la pos 3):\t{my_list}")
my_list.sort()
print(f" - Lista (ordenada):\t\t{my_list}")

# Tuple
my_tuple = (5, 3, 1, 4, 2)
print ("\n\n____ TUPLAS ____");

print(f" - Tuple:\t\t\t{my_tuple}")
print(" - Tuple: Las tuplas son inmutables, no se pueden insertar, borrar, actualizar o ordenar elementos.")

# Set
my_set = {5, 3, 1, 4, 2}
print ("\n\n____ SETS ____");

print(f" - Set:\t\t\t\t{my_set}")
my_set.add(8)
print(f" - Set (añado el 8):\t\t{my_set}")
my_set.remove(3)
print(f" - Set (borro el 3):\t\t{my_set}")

print(" - Set: los sets son conjuntos no ordenados y sin elementos repetodos, no se pueden actualizar ni ordenar elementos.")


# Dictionary
my_dict = {'j': 'value1', 'g': 'value2', 'a': 'value3'}
print ("\n\n____ DICTIONARY ____");
print(f" - Diccionario:\t\t\t\t\t{my_dict}")
my_dict['d'] = 'value4'
print(f" - Diccionario (añado el 'f'):\t\t\t{my_dict}")

my_dict.pop('a')
print(f" - Diccionario (borro el 'a' (del o pop)):\t{my_dict}")

my_dict['a'] = 'NUEVO VALOR'
print(f" - Diccionario (actualizo el 'j'):\t\t{my_dict}")

print(" - Diccionario (ordenado):")
for key in sorted(my_dict):
    print(f"\t\t\t - {key}: {my_dict[key]}")  

"""
DIFICULTAD EXTRA (opcional): AGENDA
"""
def agenda():
    
    # Clase Agenda (Dictionary)
    agenda = {}
    
    # Búsqueda de un contacto	
    def search_contact():
        name = input("Introduce el nombre del contacto a buscar: ")
        if name in agenda:
            print(f" -> Contacto encontrado. Nombre: {name} - Teléfono: {agenda.get(name)}")
        else:
            print("El contacto no existe")
    
    # Añadir/Actualizar un contacto
    def insert_contact(update = False):
        if(update):
            name_msg = "Introduce el nombre del contacto a actualizar: "
            tlfn_msg = "Introduce el nuevo teléfono del contacto: "
        else:
            name_msg = "Introduce el nombre del contacto: "
            tlfn_msg = "Introduce el teléfono del contacto: "
            
        name = input(name_msg)
        valid = False
        while(not(valid)):
            phone = input(tlfn_msg)
            if(phone.isdigit() and len(phone) == 9):
                print("Teléfono correcto")
                agenda[name] = phone
                valid = True
            else:
                retry = input(" # Teléfono no valido # ¿Desea volver a intentarlo? (s/n): ")
                if(retry.lower == "n"):
                    valid = True
                else:
                    print("Vuelva a intentarlo...")
    
    # Borrar un contacto
    def delete_contact():
        name = input("Introduce el nombre del contacto a borrar: ")
        if name in agenda:
            agenda.pop(name)
            print(f"Contacto '{name}' borrado")
        else:
            print("El contacto no existe")

    def show_agenda():
        print("Listado de contactos: ")
        for key in agenda:
            print(f"\t - {key}: {agenda[key]}")
        

    # Mostrar menú de la agenda
    def menu():
        print("\n AGENDA DE CONTACTOS\n")
        print("\t 1.- Búsqueda de un contacto")            
        print("\t 2.- Añadir un contacto")            
        print("\t 3.- Actualización de un contacto")            
        print("\t 4.- Borrar un contacto")         
        print("\t 5.- Mostrar agenda (extra)")         
        print("\t 0.- Salir")         
    
        option = input("\nSelecciona una opción: ")
        return option
    
    # Procesar la opción seleccionada
    def process_option(option):
        if(option == "1"):
            search_contact()
        elif(option == "2"):
            insert_contact()
        elif(option == "3"):
            insert_contact(True)
        elif(option == "4"):
            delete_contact()
        elif(option == "5"):
            show_agenda()
        elif(option == "0"):
            print("Saliendo de la agenda...")
        else:
            print("Opción no válida loco!")
        return option
        
    def main():
        option = ""
        while(option != "0"):
            option = menu()
            process_option(option)
        
    main()
    
agenda()

File: Roadmap/03 - ESTRUCTURAS DE DATOS/java/Worlion.java
import java.util.TreeMap;
import java.util.LinkedList;
import java.util.HashSet;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Queue;
import java.util.Scanner;

public class Worlion {

            
    Scanner scanner = new Scanner(System.in);
    HashMap<String, String> agenda = new HashMap<>();

    public static void main(String[] args) {
        System.out.println("Estructuras de datos en Java");

        new Worlion().run();
    }

    public void run() {
        estructurasDeDatos();
        extra();
    }

    /*
    * EJERCICIO: Estructuras de datos en Java
    */

    private void estructurasDeDatos() {
        // Ejemplos de creación de estructuras de datos en Java
        // Arrays
        int [] numbers = new int[5];

        // ArrayList
        ArrayList<String> listaCadenas = new ArrayList<>();
        System.out.println("Lista: " +listaCadenas);

        //HashMap
        HashMap<Integer, String> mapaEnteroCadena = new HashMap<>();
        System.out.println("Mapa: " +mapaEnteroCadena);

        //HashSet 
        HashSet<String> conjuntoCadenas = new HashSet<>();
        System.out.println("Conjunto: " +conjuntoCadenas);

        // LinkedList
        LinkedList<Integer> listaEnteros = new LinkedList<>();
        System.out.println("Lista: " +listaEnteros);

        // TreeMap
        TreeMap<String, Integer> mapaCadenaEntero = new TreeMap<>();
        System.out.println("Mapa: " +mapaCadenaEntero);

        // Cola
        Queue<Integer> cola = new LinkedList<>();
        System.out.println("Cola: " +cola);

        // inserción 
        listaCadenas.add("Hola");
        listaCadenas.add("Mundo");
        listaCadenas.add("Java");
        listaCadenas.add("Estructuras de datos");

        System.out.println("Lista tras la inserción: " +listaCadenas);
        
        //borrado, 
        listaCadenas.remove("Java");
        System.out.println("Lista tras borrado: " +listaCadenas);
        
        //actualización 
        listaCadenas.set(2, "Java");
        System.out.println("Lista tras actualización: " +listaCadenas);
        
        //ordenación.
        listaCadenas.sort(null);
        System.out.println("Lista ordenada: " +listaCadenas);
    }


    private void extra() {

        /*
        * DIFICULTAD EXTRA (opcional): Agenda de contactos
        */

        

        System.out.println(" \n ---- 🌩 DIFICULTAD EXTRA 🌩 ----\n");

        var option = "";
        while(!option.equals("0")){

            option = showMenu();
            manageOption(option);
        }
    
    }

    private String showMenu() {
        
        System.out.println("\n AGENDA DE CONTACTOS\n");
        System.out.println("\t 1.- Búsqueda de un contacto");
        System.out.println("\t 2.- Añadir un contacto");
        System.out.println("\t 3.- Actualización de un contacto");
        System.out.println("\t 4.- Borrar un contacto");
        System.out.println("\t 5.- Mostrar agenda (extra)");
        System.out.println("\t 0.- Salir");
        System.out.println("\n Selecciona una opción:");
        return scanner.nextLine();
    }

    private void manageOption(String option) {
        switch (option) {
            case "1":
                System.out.println(findContact());
                break;
            case "2":
                addContact();
                break;
            case "3":
                updateContact();
                break;
            case "4":
                deleteContact();
                break;
            case "5":
                showDirectory();
                break;
            case "0":
                System.out.println("Saliendo...");
                break;
            default:
                System.out.println("WARNING: Opción no válida");
                break;
        }
    }

    private String findContact() {
        System.out.println("Introduce el nombre del contacto que quieres buscar:");
        var name = scanner.nextLine();
        var phone = agenda.get(name);

        if(phone != null){
            return phone;
        } else {
            return "No se ha encontrado el contacto";
        }
    }

    private void addContact() {
        System.out.println("Introduce el nombre del contacto que quieres añadir:");
        var name = scanner.nextLine();
        System.out.println("Introduce el teléfono del contacto que quieres añadir:");
        var phone = readValidPhone();
        if(phone != null){
            agenda.put(name, phone);
            System.out.println("Contacto añadido");
        } else {
            System.out.println("ERROR: No se ha podido añadir el contacto");
        }
    }

    private void updateContact(){
        System.out.println("Introduce el nombre del contacto que quieres actualizar:");
        var name = scanner.nextLine();
        var phone = readValidPhone();
        if(phone != null){
            agenda.put(name, phone);
            System.out.println("Contacto actualizado");
        } else {
            System.out.println("ERROR: No se ha podido actualizar el contacto");
        }
    }

    private void deleteContact(){
        System.out.println("Introduce el nombre del contacto que quieres eliminar:");
        var name = scanner.nextLine();
        var phone = agenda.get(name);
        if(phone != null){
            agenda.remove(name);
            System.out.println("Contacto eliminado");
        } else {
            System.out.println("ERROR: No se ha encontrado el contacto");
        }
    }

    private void showDirectory(){
        System.out.println("\nAgenda de contactos:");
        for (String name : agenda.keySet()) {
            System.out.println(" - Nombre: " + name + " - Teléfono: " + agenda.get(name));
        }
        System.out.println("\nTotal de contactos: " + agenda.size() + "\n");
    }

    private String readValidPhone() {
        System.out.println("Introduce el teléfono:");
        var phone = scanner.nextLine();
        while(!validPhone(phone)){
            System.out.println(" # Teléfono no valido # ¿Desea volver a intentarlo? (s/n): ");
            var retry = scanner.nextLine();
            if(retry.equals("n")){
                return null;
            }
            System.out.println("Introduce el teléfono:");
            phone = scanner.nextLine();
        }
        return phone;
    
    }

    private boolean validPhone(String phone) {
        return phone.matches("[0-9]+") && phone.length() <= 11;
    }



}
